internal/olm/operator/internal/registry.go

```diff
@@ -0,0 +1,96 @@
+// Copyright 2019 The Operator-SDK Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package olm
+
+import (
+	"context"
+	"fmt"
+
+	olmclient "github.com/operator-framework/operator-sdk/internal/olm/client"
+
+	"github.com/operator-framework/operator-registry/pkg/registry"
+	log "github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+// SDKLabels are used to identify certain operator-sdk resources.
+var SDKLabels = map[string]string{
+	"owner": "operator-sdk",
+}
+
+// RegistryResources configures creation/deletion of internal registry-related
+// resources.
+type RegistryResources struct {
+	Client  *olmclient.Client
+	Pkg     registry.PackageManifest
+	Bundles []*registry.Bundle
+}
+
+// FEAT(estroz): allow users to specify labels for registry objects.
+
+// CreateRegistryManifests creates all registry objects required to serve
+// manifests from m.manifests in namespace.
+func (m *RegistryResources) CreateRegistryManifests(ctx context.Context, namespace string) error {
+	pkgName := m.Pkg.PackageName
+	binaryKeyValues, err := createConfigMapBinaryData(m.Pkg, m.Bundles)
+	if err != nil {
+		return fmt.Errorf("error creating registry ConfigMap binary data: %w", err)
+	}
+	cm := newRegistryConfigMap(pkgName, namespace,
+		withBinaryData(binaryKeyValues),
+	)
+	volName := getRegistryVolumeName(pkgName)
+	dep := newRegistryDeployment(pkgName, namespace,
+		withRegistryGRPCContainer(pkgName),
+		withVolumeConfigMap(volName, cm.GetName()),
+		withContainerVolumeMounts(volName, []string{containerManifestsDir}),
+	)
```

Comment:
The above impl is strange because the with would make sense it was a composition as:
`newRegistryDeployment().withA()..withB()`

However, it is not. The withs func impl here just to get the data, so wdyt about getRegistryGRPCContainer instead of for example?

Reply:
Going forward we may want to configure the registry `Deployment` differently. Using functional options allows us to configure the `Deployment` just by defining a new function to pass as an argument. The same applies to other object constructors.

As for naming see [`DynamicRESTMapperOption`](https://godoc.org/github.com/kubernetes-sigs/controller-runtime/pkg/client/apiutil#DynamicRESTMapperOption) for an example.

Reply:
I am ok with the usage of the options. 
My comment is regarding the nomenclature of the funcs .

IHMO the with would make more sense if it was used as a composition `myResourse().withA().withB()` or `buildSearch().withA().withB()` and it is not the case. 

The with funcs are creating/buiding new resources which are returned and used as param and can be used in any part of this code. So:

`deployment:= withRegistryGRPCContainer(pkgName)` (what with does?)

Would not make easier/intuitive the understatement see something as:

`deployment:= newRegistryGRPContainer(pkgName)` (what the new does?)

I hope that I could clarify my point. My suggestion here would change the name only. However, it is not a blocker.

Reply:
`deployment:= withRegistryGRPCContainer(pkgName)` will return a function with a `*v1.Deployment` parameter, not a `v1.Deployment` object. Yes I'd like to hear what others have to say as well.

Reply:
AH! :bulb:  I see what you are doing. I too thought the `withXXX` was strange but you are using it as part of a variable argument list to allow *options* to be specified when creating the Registry. That's quite clever, I never would've done it that way. I think I would've had a `Parameters` or `options` struct that I built up setting the various options and passing that in. But I kind of like the `withXXX` pattern you have going here. 

```
dep := newRegistryDeployment(pkgName, namespace,
		withRegistryGRPCContainer(pkgName),
		withVolumeConfigMap(volName, cm.GetName()),
		withContainerVolumeMounts(volName, []string{containerManifestsDir}),
```
Considering the above code from your example, I read this as create me a new RegistryDeployment using a container with GRPC, with a volume configmap with a list of volumemounts.

Again, I never would've approached it that way but it met my time criteria, that is *how long does it take me to figure out what this function is doing*. I'm okay with the current implementation.

Reply:
@jmrodri Yeah the use of [functional options](https://github.com/tmrts/go-patterns/blob/master/idiom/functional-options.md) is a common pattern that we've used before in the SDK, and upstream in the controller-runtime client and logger.
Though [method chaining](https://github.com/tmrts/go-patterns/blob/master/creational/builder.md) is probably easier to understand at first glance there are benefits to using functional options when returning errors.
Neither are idiomatic in Go but I prefer functional options.

@estroz I do think we can probably clean this up a bit to actually define Options structs and type aliases for the functional option types for both the RegistryConfigMap options and the RegistryDeployment options.
Take a look at how the controller-runtime's logger defines it's functional options.
https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/log/zap/zap.go

We can have an Options struct for just the RegistryDeployment options and have the helpers only set the options. e.g(roughly speaking):

```Go
type Opts func(*RegistryDeploymentOptions)

type RegistryDeploymentOptions struct {
    PkgName string
    ...
}

func withRegistryGRPCContainer(pkgName string) Opts {
    return func(o *Options) {
		o.PkgName = pkgName
	}
}
```
And then `newRegistryDeployment()` just applies all the `Opts` passed in to a default RegistryDeploymentOptions struct and then creates the deployment based on those options. E.g how the logger constructor does it:
https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/log/zap/zap.go#L195-L209